index.rst

```diff
@@ -1,29 +1,375 @@
+The primary advantage of HTTP headers is to optimize away the API call to Gafaelfawr and the extra overhead in creating a delegated token, but we are avoiding premature optimization until we have evidence it is a problem.
+
+There are many ratelimiting packages available in Python.
+We chose limits_ because it supports Redis, asyncio, and the type of configuration that's required for use inside Gafaelfawr.
+It also supports a wide variety of rate limit algorithms if we want to change fixed-window to something more sophisticated.
+It unfortunately depends on a different Redis library (maintained by the same author), so this introduces a second Redis library into our infrastructure, but the other advantages outweighed this.
+
+Other options considered:
+
+- `fastapi-limiter <https://github.com/long2ice/fastapi-limiter>`__ wants to be invoked as a FastAPI dependency.
+  This is great for rate limiting within a FastAPI application, and we should consider it again when we need to move rate limiting into the individual service, but it wants to run as a dependency and relies on being able to extract the route from the request.
+  All Gafaelfawr rate limit checking happens in the ``/auth`` route, and Gafaelfawr needs to be able to rate limit on the basis of the user and service extracted from the token.
+
+- `ASGI RateLimit <https://github.com/abersheeran/asgi-ratelimit>`__ has a similar problem: it wants to get all the configuration for the rate limiting and applies it by analyzing the incoming route.
+
+- `aiolimiter <https://aiolimiter.readthedocs.io/en/latest/>`__ and `SlowApi <https://slowapi.readthedocs.io/en/latest/>`__ only work in-memory in a single process and don't support a shared rate limit in Redis.
+
+- `python-redis-rate-limit <https://github.com/EvoluxBR/python-redis-rate-limit>`__ has a good API and the most sophisticated counter implementation using the Lua script recommended by the `Redis documentation <https://redis.io/commands/incr/#pattern-rate-limiter-2>`__.
+  Unfortunately, it doesn't support asyncio, which is a requirement for Gafaelfawr.
+  (It also has a 0.0.8 version number.)
+
+- `asyncio-redis-rate-limit <https://github.com/wemake-services/asyncio-redis-rate-limit>`__ has all the required features, but the key generation algorithm seems dodgy to me and it uses a relatively unsophisticated fixed-window algorithm.
+
+Finally, NGINX can do rate limiting directly.
+This can be configured per-ingress with `annotations <https://kubernetes.github.io/ingress-nginx/user-guide/nginx-configuration/annotations/#rate-limiting>`__.
+Using this rate limiting would be the least effort for us.
+
+However, since NGINX has no access to the user's authentication information, it cannot do rate-limiting by user, only by IP address.
+Since we expect many requests to come from inside the cluster via other services such as the Portal or Notebook Aspect, this cannot be used for the type of rate limiting we want to do.
```

Comment:
Also, of course, many institutions have NATted traffic that appear to come from a single IP address, and we can't rate limit all of Princeton just because Robert's hammering the service.